content/blog/2019-09-29-tbaa.md

```diff
@@ -0,0 +1,180 @@
+if there is no alias inside the loop. Similar optimizations also apply to branches. 
+
+#### Results
+By statically examining the number of removed redundant loads, we can draw the same conclusion as in the static evaluation 
+section: *FieldTypeDecl* and *SMFieldTypeRefs* are strictly more powerful than *TypeDecl*. However, when comparing the 
+performance of optimized programs, all three techniques offer similar speedup when used together with RLE. This result 
+demonstrates that at least for RLE, a more precise alias analysis may not provide much benefit over TBAA. The limit analysis
+confirms this statement, because almost all redundant loads can be removed by using RLE together with TBAA for most benchmarks. 
+For the redundant loads that can not be removed, the authors studied the cause and found that they are mostly caused by problems
+other than alias analysis. 
+
+### Performance on Incomplete Programs
+The assumption of the compiler having access to the whole program is often violated in cases such as separate compilation. 
+As a result, the authors performed additional experiments to evaluate how TBAA performs when this assumption does not hold. 
+The authors use "open world" to refer to the cases where the assumption does not hold, and use "close world" to refer to 
+situations where the whole program is available to the compiler. To ensure that TBAA on incomplete programs still yield 
+correct results in the "open world" scenario, the authors made some changes to make it more conservative:
+- When evaluating whether the address of a memory access has been taken elsewhere, the modified version also checks the 
+function arguments. If two function arguments are both references and have the same type, then the instructions accessing
+these two function arguments might alias, because the code that calls the function might assign aliasing objects to the function arguments. 
+- When merging types in *SMFieldTypeRefs*, any two types with a subtype relationship are merged together, since the 
+unavailable code might assign them. To my understanding, this modification makes *SMFieldTypeRefs* almost equivalent to 
+*FieldTypeDecl*. 
+
+While these changes seem very conservative, actually they do not affect the performance of TBAA that much. The authors show 
+that the "open world" assumption has negligible effect on the execution time of the compiled program when RLE is applied. The 
+result is not surprising, because even in the "close world" scenario, *SMFieldTypeRefs* does not have clear advantage over
+*FieldTypeDecl*. The modified version of *SMFieldTypeRefs* is very close to *FieldTypeDecl*, with just slightly more 
+restrictions. However, it is not mentioned in the paper **how much** of the program is still available in the "open world" 
+case, which makes the result less convincing. TBAA with only one function available and with all but one functions available
```

Comment:
While I don't know for sure, I assume the answer to "how much" of the code is available is: all of it. That is, the difference is not in how much code the analysis gets to see but in whether it needs to assume that there *might* be additional code out there that it doesn't get to observe.